modules/core/src/shaderlib/shadow/shadow.js

```diff
@@ -0,0 +1,186 @@
+      }
+    }
+  }
+  return gl_Position;
+}
+`;
+
+const fs = `
+const int max_lights = 2;
+uniform bool shadow_uDrawShadowMap;
+uniform bool shadow_uUseShadowMap;
+uniform sampler2D shadow_uShadowMap[max_lights];
+uniform vec4 shadow_uColor;
+uniform float shadow_uLightCount;
+
+varying vec3 shadow_vPosition[max_lights];
+
+const vec4 bitPackShift = vec4(1.0, 255.0, 65025.0, 16581375.0);
+const vec4 bitUnpackShift = 1.0 / bitPackShift;
+const vec4 bitMask = vec4(1.0 / 255.0, 1.0 / 255.0, 1.0 / 255.0,  0.0);
+
+float shadow_getShadowWeight(vec3 position, sampler2D shadowMap) {
+  vec4 rgbaDepth = texture2D(shadowMap, position.xy);
+
+  float z = dot(rgbaDepth, bitUnpackShift);
+  return smoothstep(0.001, 0.01, position.z - z);
+}
+
+vec4 shadow_filterShadowColor(vec4 color) {
+  if (shadow_uDrawShadowMap) {
+    vec4 rgbaDepth = fract(gl_FragCoord.z * bitPackShift);
+    rgbaDepth -= rgbaDepth.gbaa * bitMask;
+    return rgbaDepth;
+  }
+  if (shadow_uUseShadowMap) {
+    float shadowAlpha = 0.0;
+    for (int i = 0; i < max_lights; i++) {
+      if(i < int(shadow_uLightCount)) {
+        shadowAlpha += shadow_getShadowWeight(shadow_vPosition[i], shadow_uShadowMap[i]) * shadow_uColor.a / 255.0 / shadow_uLightCount;
+      }
+    }
+    float blendedAlpha = shadowAlpha + color.a * (1.0 - shadowAlpha);
+
+    return vec4(
+      mix(color.rgb, shadow_uColor.rgb / 255.0, shadowAlpha / blendedAlpha),
+      blendedAlpha
+    );
+  }
+  return color;
+}
+`;
+
+const moduleName = 'shadow';
+
+createModuleInjection(moduleName, {
+  hook: 'vs:DECKGL_FILTER_GL_POSITION',
+  injection: `
+position = shadow_setVertexPosition(geometry.position);
+  `
+});
+
+createModuleInjection(moduleName, {
+  hook: 'fs:DECKGL_FILTER_COLOR',
+  injection: `
+color = shadow_filterShadowColor(color);
+  `
+});
+
+const DEFAULT_shadow_Color = [0, 0, 0, 255];
+const VECTOR_TO_POINT_MATRIX = [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0];
+
+function createShadowUniforms(opts = {}, context = {}) {
+  const uniforms = {
+    shadow_uDrawShadowMap: Boolean(opts.drawToShadowMap),
+    shadow_uUseShadowMap: opts.shadowMaps ? opts.shadowMaps.length > 0 : false,
+    shadow_uColor: opts.shadowColor || DEFAULT_shadow_Color,
+    shadow_uLightId: opts.shadow_lightId,
+    shadow_uLightCount: opts.shadow_viewProjectionMatrices.length
+  };
+
+  const center = new Matrix4(opts.viewport.viewProjectionMatrix)
+    .invert()
+    .transformVector4(context.project_uCenter);
+
+  const viewProjectionMatrices = [];
+  const projectCenters = [];
+
+  for (let i = 0; i < opts.shadow_viewProjectionMatrices.length; i++) {
+    const viewProjectionMatrix = opts.shadow_viewProjectionMatrices[i]
+      .clone()
+      .translate(new Vector3(opts.viewport.center).negate());
+
+    if (context.project_uCoordinateSystem === PROJECT_COORDINATE_SYSTEM.LNG_LAT) {
+      viewProjectionMatrices.push(viewProjectionMatrix);
+      projectCenters.push([0, 0, 0, 0]);
```

Comment:
Use `array[i]` for assignment since you are targeting a specific uniform. `push` is less readable and prone to bugs.